---
title: "rna_seq_notes_21"
author: "lau-lhg"
date: "2/23/2021"
output: html_document
---
Original notes taken in notion.so 

```{r eval=FALSE}
## Opción más nueva:
library("gert")
repo <- git_clone(
    "https://github.com/lcolladotor/rnaseq_LCG-UNAM_2021",
    "~/rnaseq_LCG-UNAM_2021"
)
setwd(repo)

```

# RNA-seq

Class guide: [https://lcolladotor.github.io/rnaseq_LCG-UNAM_2021/introducción-a-r-y-rstudio.html](https://lcolladotor.github.io/rnaseq_LCG-UNAM_2021/introducci%C3%B3n-a-r-y-rstudio.html)

▶️ This class will focus on RNA-seq data analysis

If you don't want to install all the libraries, you can use the server: [http://132.248.220.108:8787/](http://132.248.220.108:8787/)

---

### Creating an R project and linking to GitHub

- Exercise 1: My first R project (this repository)

From R studio, you can create a project. Make sure you're standing in whichever directory you want to build it or use the absolute route in the argument. 

```{r eval=FALSE}
#create project
usethis::create_project("/path/my_new_project")
##setup file
usethis::use_r("name.R")
#once your project is created, you can add .r and .rmd files within it
```

To connect it to GitHub, you must first obtain a token to validate your identity. Run the line below and a pop-up will open, from where you must fill out your name (once you're signed in to GitHub) and then press the large green button at the bottom. Save the token you are given.

```{r eval=FALSE}
#get token
usethis::create_github_token()
#then this
gitcreds::gitcreds_set()
#you will be asked for your token here

```

While using the server, something with the token went wrong. An alternative way to configure it is: 

```{r eval=FALSE}
usethis::edit_r_environ()
#This will open a file
```

Within the file write your token and add a new line, save file, and restart R session.

```{r eval=FALSE}
GITHUB_PAT= token

```

Now we can finish the GitHub configuration:

```{r eval=FALSE}
#this will open another file
usethis::edit_git_config()
#fill it out with your GitHub account info

#connect
usethis::use_git()
usethis::use_github()
```

- Postcard Exercise

    Create your own profile/webpage (git was already configured)

    ```{r eval=FALSE}
    #create project
    usethis::create_project("lau-lhg.github.io")

    #start git
    usethis::use_git()
    usethis::use_github()

    #postcard
    postcards::create_postcard(template = "onofre")
    #other templates are: "jolla", "jolla-blue", "trestles"

    #then commit and push channges
    ```

    Result: 
[https://lau-lhg.github.io/](https://lau-lhg.github.io/)

### Bio-conductor and Packages

> "Analysis and comprehension of high-throughput genomic data"

---

Packages can be classified as either: software, annotation, experiment, workflow.

[Bioconductor Software Packages](http://bioconductor.org/packages/3.12/bioc/)

There's a whole community dedicated to testing and improving packages, but it's important to be mindful of the code you use, to know if using it is a good idea.

- Team Exercise: Select one or two packages each and discuss.
**Team: Ana Sofia Escobedo, Leonardo Correa, Laura Hinojosa**

My Package of choice: spatialHeatMap

I liked that it had been widely tested.The code seems well documented, and is open-source (duh). there were no issues reported in GitHub. I personally enjoy working with graphs and figures, which is mostly the reason for my preferring this, but it looks that the team behind this package is quite active and using it sounds like a decent idea.

### Summarized Experiment
Summarized experiment es un objeto cuya parte principal consiste de tres tablas unidas. 

- rowRanges: tabla de información de los genes, un renglón por cada gen. Es un objeto tipo genomic ranges.
- assays: (one or more) un renglon por gen, una columns por muestra.
- colData: un renglón por cada muestra, una columna por cada variable de info.

Es posible crear versiones distintas de este tipo de dato, con más tablas relacionadas. Es fácil conceptualizarlo como una base de datos.

rtrach layer es un paquete que te permite leer datos de distintos formatos bioinformáticos (BED, GTF, etc).

```{r}
## Lets build our first SummarizedExperiment object
library("SummarizedExperiment")
## ?SummarizedExperiment

## De los ejemplos en la ayuda oficial

## Creamos los datos para nuestro objeto de tipo SummarizedExperiment
## para 200 genes a lo largo de 6 muestras
nrows <- 200
ncols <- 6
## Números al azar de cuentas
set.seed(20210223)
counts <- matrix(runif(nrows * ncols, 1, 1e4), nrows)
## Información de nuestros genes
rowRanges <- GRanges(
    rep(c("chr1", "chr2"), c(50, 150)),
    IRanges(floor(runif(200, 1e5, 1e6)), width = 100),
    strand = sample(c("+", "-"), 200, TRUE),
    feature_id = sprintf("ID%03d", 1:200)
)
names(rowRanges) <- paste0("gene_", seq_len(length(rowRanges)))
## Información de nuestras muestras
colData <- DataFrame(
    Treatment = rep(c("ChIP", "Input"), 3),
    row.names = LETTERS[1:6]
)
## Juntamos ahora toda la información en un solo objeto de R
rse <- SummarizedExperiment(
    assays = SimpleList(counts = counts),
    rowRanges = rowRanges,
    colData = colData
)

## Exploremos el objeto resultante
rse
```

```{r}
#tabla assay
head(assay(rse))
```
```{r}
#ver los cromosomas
seqlevels(rse)
```
```{r}
#Ejercicio

## Comando 1
rse[1:2, ]
head(assay(rse[1:2, ]))

```
Como los renglones corresponden a los genes, aqui nos muestra un sub-set con sólo los primeros 2 genes.

```{r}
## Comando 2
rse[, c("A", "D", "F")]
head(assay(rse[, c("A", "D", "F")]))
#tras bambalinas
which(colnames(rse)%in%c("A", "D", "F"))
```
Como las columnas son muestras, aquí accesamos a un subset con las primeras tres.

```{r}
#tambien podemos accesar así
rse$Treatment
```

Exploremos las tres sub-tablas de nuestro objeto.
```{r}
print('ASSAY')
head(assay(rse))
print('ROWRANGES')
head(rowRanges(rse))
print('COLDATA')
head(colData(rse))
```
To view all of this information in painfully beautiful graphs, do:
```{r eval=FALSE}
## Explora el objeto rse de forma interactiva
library("iSEE")
iSEE::iSEE(rse)
```

#### Exercise 2
```{r}
## Descarguemos unos datos de spatialLIBD
sce_layer <- spatialLIBD::fetch_data("sce_layer")

## Revisemos el tamaño de este objeto
pryr::object_size(sce_layer)


```
Generate first graph
```{r eval=FALSE}
iSEE::iSEE(sce_layer)
```

[image 1]

Obtain a shortened version of the genes we're interested in.
```{r}
short<-sce_layer[which(rowRanges(sce_layer)$gene_name%in%c("MOBP","MBP","PCP4")),]
```
Clustered heatmap of subset
```{r eval=FALSE}
iSEE::iSEE(short)

```

[image 2]

### Recount

Se limitó a humano y ratón porque son los organismos con mayor estudios de RNAseq.
```{r}
# Load recount3 R package
library("recount3")
human_projects <- available_projects()
``` 

Tomamos el id del experimento de interés y se crea un objeto.

```{r eval=FALSE}
proj_info <- subset(
    human_projects,
    project == "SRP009615" & project_type == "data_sources"
)
## Crea un objetio de tipo RangedSummarizedExperiment (RSE)
## con la información a nivel de genes
rse_gene_SRP009615 <- create_rse(proj_info)
rse_gene_SRP009615
```
```{r eval=FALSE}
## Explora los proyectos disponibles de forma interactiva
proj_info_interactive <- interactiveDisplayBase::display(human_projects)
## Selecciona un solo renglón en la tabla y da click en "send".

## Aquí verificamos que solo seleccionaste un solo renglón.
stopifnot(nrow(proj_info_interactive) == 1)
## Crea el objeto RSE
rse_gene_interactive <- create_rse(proj_info_interactive)

```

```{r eval=FALSE}
assay(rse_gene_SRP009615, "counts") <- compute_read_counts(rse_gene_SRP009615)

rse_gene_SRP009615 <- expand_sra_attributes(rse_gene_SRP009615)
colData(rse_gene_SRP009615)[
    ,
    grepl("^sra_attribute", colnames(colData(rse_gene_SRP009615)))
]
```

La función de expand sólo se puede utilizar si sólo hay un set de datos. 

#### Ejercicio
```{r eval=FALSE}

iSEE::iSEE(rse_gene_SRP009615)
```

[image 3]

